# Evaluating Pre-trained model for closed QnA

### Please refer to the respective sections in the book for further details.


In [140]:
from datasets import load_dataset

In [155]:
dataset = load_dataset('pubmed_qa', 'pqa_labeled')
dataset

  0%|          | 0/1 [00:00<?, ?it/s]

DatasetDict({
    train: Dataset({
        features: ['pubid', 'question', 'context', 'long_answer', 'final_decision'],
        num_rows: 1000
    })
})

In [156]:
# dataset = load_dataset('duorc', 'SelfRC')
# dataset

## Extract 1000 questions for evaluation

In [159]:
# Extract 10 examples from the dataset
train_dataset = dataset['train'].shuffle().select(range(800))

test_dataset = dataset['train'].shuffle().select(range(800, 1000))

# Extract questions, answers, and context (plot)
questions = test_dataset['question']
answers = test_dataset['long_answer']

# Extract 'contexts' from the 'context' column
contexts = test_dataset['context']

# Convert the 'contexts' list to strings for each row
context_strings = [' '.join(context['contexts']) for context in contexts]

contexts_list = []
# Print the extracted and converted context strings
for context_string in context_strings:
    contexts_list.append(context_string)
#     print("Context:", context_string)

# Print the first question, answer, and context
print("Question:", questions[9])
print("Answer:", answers[9])
print("Context:", contexts_list[9])


Question: Do socioeconomic differences in mortality persist after retirement?
Answer: Socioeconomic differences in mortality persist beyond retirement age and in magnitude increase with age. Social differentials in mortality based on an occupational status measure seem to decrease to a greater degree after retirement than those based on a non-work measure. This suggests that alongside other socioeconomic factors work itself may play an important part in generating social inequalities in health in men of working age.
Context: To assess the risk of death associated with work based and non-work based measures of socioeconomic status before and after retirement age. Follow up study of mortality in relation to employment grade and car ownership over 25 years. The first Whitehall study. 18,133 male civil servants aged 40-69 years who attended a screening examination between 1967 and 1970. Death. Grade of employment was a strong predictor of mortality before retirement. For men dying at ages 

## Evaluating pre-trained model



### Load model locally

### Generating answer from HuggingFace model

In [161]:
import requests
import pandas as pd
import json

# API URL for the Hugging Face model
# API_URL = "https://api-inference.huggingface.co/models/gpt2"
API_URL = "https://api-inference.huggingface.co/models/bigscience/bloom"
# API_URL = "https://api-inference.huggingface.co/models/openlm-research/open_llama_13b"
# API_URL = "https://api-inference.huggingface.co/models/EleutherAI/gpt-neox-20b"
# API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xxl"
# API_URL = "https://api-inference.huggingface.co/models/tiiuae/falcon-7b-instruct"

# Create an empty list to store the results
results = []

# Iterate over each question, answer, and context
for idx, (question, answer, context) in enumerate(zip(questions, answers, contexts_list)):
#     print(question, answer)
#     print("--------------context------------", context[:700])
    # Create the model input prompt
    model_input = f"Answer the question based on the context below.\nContext: {context}\nQuestion: {question}"
    headers = {"Authorization": "Bearer XXXXXXXXXXXXXXXXXX"} ## ADD YOUR BEARER TOKEN
#     print("=========model_input=========", model_input)
    # Define the payload for the API request
    payload = {
        "inputs": model_input,
        "parameters": {'temperature': 0.5,
                    'max_new_tokens': 50,
                    'return_full_text': False,
#                    'truncation': 'only_first'
                    },
    }

    # HuggingFace response
    response = requests.post(API_URL, headers=headers, json=payload)
    json_response = json.loads(response.content.decode("utf-8"))
    print("*********json_response**********",json_response)
    model_output = json_response[0]['generated_text']

    # Append the results to the list
    results.append({
        'Question': question,
        'Actual Answer': answer,
        'Generated Answer': model_output,
        'Context': context
    })

    # Print progress
    print(f"Processed {idx+1} / {len(questions)}")

# Create a Pandas DataFrame from the results
df = pd.DataFrame(results)

# Print the DataFrame
df.head()


*********json_response********** [{'generated_text': " Can it serve as a surrogate marker of cardiovascular risk among black Africans?\nI have used logistic regression and Pearson's correlation coefficient to determine the relationship between these variables but I do not know if this is the correct procedure.\nPlease help me with this"}]
Processed 1 / 200
*********json_response********** [{'generated_text': ' The authors conclude that there is no evidence for intrauterine effects on obesity. They base this conclusion on the similar association of maternal and paternal BMI with offspring BMI. The authors conclude that there is no evidence for intrauterine effects on obesity. They'}]
Processed 2 / 200
*********json_response********** [{'generated_text': ' And does it prevent bacterial colonization in vivo?'}]
Processed 3 / 200
*********json_response********** [{'generated_text': ''}]
Processed 4 / 200
*********json_response********** [{'generated_text': ' Please explain your answer with

*********json_response********** [{'generated_text': ''}]
Processed 41 / 200
*********json_response********** [{'generated_text': ''}]
Processed 42 / 200
*********json_response********** [{'generated_text': ' If so, how much improvement has occurred over time?\nThe question is whether survival is improving over time. The authors found a statistically significant improvement in survival across five groups of patients, and the more recent groups had longer survival. The question is whether this'}]
Processed 43 / 200
*********json_response********** [{'generated_text': ' Answer: No.'}]
Processed 44 / 200
*********json_response********** [{'generated_text': ''}]
Processed 45 / 200
*********json_response********** [{'generated_text': ' What is the incidence of anastomotic leakage in this group of patients?\nThe conclusion of the article is that rectal anastomosis in gynaecological debulking surgery without a diverting stoma is safe. The incidence of'}]
Processed 46 / 200
*********json_respo

*********json_response********** [{'generated_text': ''}]
Processed 75 / 200
*********json_response********** [{'generated_text': ' Do you agree with the conclusion of the authors? What could be the explanation for the low prevalence of severe ROP in the SGA group?\nDiscussion:\nIn this cohort study, the authors found that being small for gestational age was not'}]
Processed 76 / 200
*********json_response********** [{'generated_text': ' What do you think is the reason for their lack of knowledge?\nAnswer:\nThe correct answer is B. An incorrect answer is A. The correct answer is C. An incorrect answer is D.'}]
Processed 77 / 200
*********json_response********** [{'generated_text': ' If so, does the accuracy depend on the sampling device or on the diagnosis of vaginal infections?\nAnswer: The vaginal pH can be measured from the wet mount slide. However, the accuracy is low and depends on the sampling device and on the'}]
Processed 78 / 200
*********json_response********** [{'generated_t

*********json_response********** [{'generated_text': ' Explain.\nAnswer: Quality of life in lung cancer patients: does socioeconomic status matter?\nExplain.'}]
Processed 112 / 200
*********json_response********** [{'generated_text': ''}]
Processed 113 / 200
*********json_response********** [{'generated_text': ' What is the therapeutic yield of this method? What are the advantages and disadvantages of this method?\nThe aim of this study was to analyze the contribution of the double-balloon enteroscopy (DBE) for diagnosis of the small bowel disorders'}]
Processed 114 / 200
*********json_response********** [{'generated_text': ' If yes, what is the optimal set of measures associated with regional plantar surface area?\nA. Yes, a two variable model was found to describe the relationship between the foot measurements and forefoot plantar contact area (r(2)=0.79,'}]
Processed 115 / 200
*********json_response********** [{'generated_text': '"\n\nA:\n\nI take it that the question is not asking 

*********json_response********** [{'generated_text': ''}]
Processed 157 / 200
*********json_response********** [{'generated_text': ''}]
Processed 158 / 200
*********json_response********** [{'generated_text': ' If so, how much? What is the effect of police enforcement on traffic fatalities and severe injuries?\nThe new traffic law in Chile introduced in 2005, which has been described in detail in previous studies (1, 2) , has been a milestone'}]
Processed 159 / 200
*********json_response********** [{'generated_text': ''}]
Processed 160 / 200
*********json_response********** [{'generated_text': " If so, what is the relative strength of the predictive value of scores on a preclerkship clinical performance examination and scores on the United States Medical Licensing Examination (USMLE) Step 1?\nDiscussion: The authors found that students' scores"}]
Processed 161 / 200
*********json_response********** [{'generated_text': ''}]
Processed 162 / 200
*********json_response********** [{'generat

*********json_response********** [{'generated_text': ' If so, how much do they predict?\nAnswer: Yes, somatic complaints predict subsequent symptoms of depression, but not as much as income and age.\nComments: This is a very well designed study. The authors are to be congratulated for their'}]
Processed 200 / 200


Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer  \
0  Our findings support CIMT as a significant ind...   
1  Comparison of mother-offspring and father-offs...   
2  Implant coating with an antibacterial-loaded h...   
3  This study that shows that adequate femoral tu...   
4  The preconditioning effect (infarct size reduc...   

                                    Generated Answer  \
0   Can it serve as a surrogate marker of cardiov...   
1   The authors conclude that there is no evidenc...   
2   And does it prevent bacterial colonization in...   
3                                                      
4   Please explain your answer with the context o...   

                                             Context  
0  It is not known whether common carotid intima ...  
1  It has been suggested that increasing obesity ...  
2  Implant-related infections represent one of th...  
3  This study aims to study femoral tunnel length...  
4  Ischemic preconditioning (IP) is initiated thr...

## Evaluation metrics



In [152]:
# !pip install bert-score
# !pip install nltk
# !pip install rouge-score
# !pip install sentence-transformers


In [185]:
from bert_score import score as bert_score
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge import Rouge
from sentence_transformers import SentenceTransformer
import numpy as np
import re
import string
import collections
from rouge_score import rouge_scorer

def normalize_answer(s):
  """Lower text and remove punctuation, articles and extra whitespace."""
  def remove_articles(text):
    regex = re.compile(r'\b(a|an|the)\b', re.UNICODE)
    return re.sub(regex, ' ', text)
  def white_space_fix(text):
    return ' '.join(text.split())
  def remove_punc(text):
    exclude = set(string.punctuation)
    return ''.join(ch for ch in text if ch not in exclude)
  def lower(text):
    return text.lower()
  return white_space_fix(remove_articles(remove_punc(lower(s))))

def get_tokens(s):
    if not s: return []
    return normalize_answer(s).split()

# Compute F1 score
def compute_f1(a_gold, a_pred):
    gold_toks = get_tokens(a_gold)
    pred_toks = get_tokens(a_pred)
    common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
    num_same = sum(common.values())
    if len(gold_toks) == 0 or len(pred_toks) == 0:
        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
        return int(gold_toks == pred_toks)
    if num_same == 0:
        return 0
    precision = 1.0 * num_same / len(pred_toks)
    recall = 1.0 * num_same / len(gold_toks)
    f1 = (2 * precision * recall) / (precision + recall)
    return f1

# Calculate F1 score
def calculate_f1_score(row):
    return compute_f1(row['Actual Answer'], row['Generated Answer'])

# Calculate BLEU score
def calculate_bleu_score(row):
    smoothie = SmoothingFunction().method4
    reference = row['Actual Answer'].split()
    hypothesis = row['Generated Answer'].split()
    return sentence_bleu([reference], hypothesis, smoothing_function=smoothie)

# Compute ROUGE score
def compute_rouge(answer, ideal_answer):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    # Calculate the ROUGE score
    score = scorer.score(answer, ideal_answer)
    # Extract the F1 score for ROUGE-1
    rouge_score = score['rouge1'].fmeasure
    return rouge_score

# Calculate ROUGE score
def calculate_rouge_score(row):
    return compute_rouge(row['Actual Answer'], row['Generated Answer'])

# Calculate SentenceSim score
def calculate_sentencesim_score(row):
    model = SentenceTransformer('average_word_embeddings_glove.6B.300d')
    embeddings1 = model.encode(row['Generated Answer'])
    embeddings2 = model.encode(row['Actual Answer'])
    similarity = np.dot(embeddings1, embeddings2) / (np.linalg.norm(embeddings1) * np.linalg.norm(embeddings2))
    return similarity

# Apply the functions to the DataFrame
df['F1 Score'] = df.apply(calculate_f1_score, axis=1)
df['BLEU Score'] = df.apply(calculate_bleu_score, axis=1)
df['ROUGE Score'] = df.apply(calculate_rouge_score, axis=1)
df['SentenceSim Score'] = df.apply(calculate_sentencesim_score, axis=1)

# Print the DataFrame with evaluation metrics
print(df.head())
df.describe()

/var/folders/gy/6lyfn5752472j2htx4g01gc40000gn/T/ipykernel_40154/2903860368.py:84: RuntimeWarning: invalid value encountered in float_scalars
  similarity = np.dot(embeddings1, embeddings2) / (np.linalg.norm(embeddings1) * np.linalg.norm(embeddings2))


                                            Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer  \
0  Our findings support CIMT as a significant ind...   
1  Comparison of mother-offspring and father-offs...   
2  Implant coating with an antibacterial-loaded h...   
3  This study that shows that adequate femoral tu...   
4  The preconditioning effect (infarct size reduc...   

                                    Generated Answer  \
0   Can it serve as a surrogate marker of cardiov...   
1   The authors conclude that there is no evidenc...   
2   And does it prevent bacterial colonization in...   
3                                                      
4   Please explain your answer with the contex

BLEU Score alternate    F1 Score  BLEU Score  ROUGE Score  \
count          2.000000e+02  200.000000  200.000000   200.000000   
mean           1.018309e-02    0.130776    0.017924     0.154657   
std            3.391808e-02    0.133483    0.032174     0.150328   
min            0.000000e+00    0.000000    0.000000     0.000000   
25%            0.000000e+00    0.000000    0.000000     0.000000   
50%           9.748507e-232    0.120606    0.008375     0.155048   
75%           5.209998e-155    0.236741    0.022213     0.282472   
max            2.576455e-01    0.500000    0.262998     0.500000   

       SentenceSim Score  
count         120.000000  
mean            0.691567  
std             0.176253  
min             0.057806  
25%             0.627781  
50%             0.740410  
75%             0.822781  
max             0.934191

#### Observation
Scores for the BLOOM model on 200 QnA test dataset:
- F-1 score: 0.133483	
- BLEU score: 0.032174
- ROUGE Score: 0.150328
- SentenceSim score: 0.176253

### Alternate BLUE and ROGUE calculation
#### Calculation validation

In [188]:
import nltk
from sentence_transformers import SentenceTransformer, util
model1 = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

def compute_blue_alternate(answer, ideal_answer):
    # Tokenize the generated and reference answers
    generated_tokens = nltk.word_tokenize(answer)
    reference_token_lists = [nltk.word_tokenize(answer) for answer in [ideal_answer]]
    # Calculate the BLEU score
    bleu_score =  nltk.translate.bleu_score.sentence_bleu(reference_token_lists, generated_tokens)
    return bleu_score

# Calculate F1 score
def calculate_f1_alternate(row):
    return compute_blue_alternate(row['Actual Answer'], row['Generated Answer'])

def compute_rouge_alternate(answer, ideal_answer):

    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    # Calculate the ROUGE score
    score = scorer.score(answer, ideal_answer)
    # Extract the F1 score for ROUGE-1
    rouge_score = score['rouge1'].fmeasure
    return rouge_score

# Calculate rouge score
def calculate_rouge_alternate(row):
    return compute_rouge_alternate(row['Actual Answer'], row['Generated Answer'])

def sentence_similarity_alternate(ideal_answer, generated_answer):
    #Compute embedding for both lists
    embedding_1= model1.encode(ideal_answer, convert_to_tensor=True)
    embedding_2 = model1.encode(generated_answer, convert_to_tensor=True)
    sim_score = util.pytorch_cos_sim(embedding_1, embedding_2)
    sim_score = sim_score.numpy()[0][0]
    return sim_score

# Calculate rouge score
def calculate_similarity_alternate(row):
    return sentence_similarity_alternate(row['Actual Answer'], row['Generated Answer'])

df['BLEU Score alternate'] = df.apply(calculate_f1_alternate, axis=1)
df['Rouge Score alternate'] = df.apply(calculate_rouge_alternate, axis=1)
df['Similarity Score alternate'] = df.apply(calculate_similarity_alternate, axis=1)

df.describe()

/Users/shivamsolanki/anaconda3/lib/python3.10/site-packages/nltk/translate/bleu_score.py:552: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/Users/shivamsolanki/anaconda3/lib/python3.10/site-packages/nltk/translate/bleu_score.py:552: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/Users/shivamsolanki/anaconda3/lib/python3.10/site-packages/nltk/translate/bleu_score.py:552: UserWarning: 
The hypothesis contains 0 counts of 2-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using low

BLEU Score alternate    F1 Score  BLEU Score  ROUGE Score  \
count          2.000000e+02  200.000000  200.000000   200.000000   
mean           1.018309e-02    0.130776    0.017924     0.154657   
std            3.391808e-02    0.133483    0.032174     0.150328   
min            0.000000e+00    0.000000    0.000000     0.000000   
25%            0.000000e+00    0.000000    0.000000     0.000000   
50%           9.748507e-232    0.120606    0.008375     0.155048   
75%           5.209998e-155    0.236741    0.022213     0.282472   
max            2.576455e-01    0.500000    0.262998     0.500000   

       SentenceSim Score  Rouge Score alternate  Similarity Score alternate  
count         120.000000             200.000000                  200.000000  
mean            0.691567               0.154657                    0.328476  
std             0.176253               0.150328                    0.324277  
min             0.057806               0.000000                   -0.114504  
25%             0.627781               0.000000                    0.004425  
50%             0.740410               0.155048                    0.374280  
75%             0.822781               0.282472                    0.647597  
max             0.934191               0.500000                    0.889436

#### Similarity score alternate looks more accurate

In [184]:
df.head()

Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer  \
0  Our findings support CIMT as a significant ind...   
1  Comparison of mother-offspring and father-offs...   
2  Implant coating with an antibacterial-loaded h...   
3  This study that shows that adequate femoral tu...   
4  The preconditioning effect (infarct size reduc...   

                                    Generated Answer  \
0   Can it serve as a surrogate marker of cardiov...   
1   The authors conclude that there is no evidenc...   
2   And does it prevent bacterial colonization in...   
3                                                      
4   Please explain your answer with the context o...   

                                             Context  BLEU Score alternate  
0  It is not known whether common carotid intima ...         3.863897e-155  
1  It has been suggested that increasing obesity ...         1.179890e-231  
2  Implant-related infections represent one of th...         5.157007e-155  
3  This study aims to study femoral tunnel length...          0.000000e+00  
4  Ischemic preconditioning (IP) is initiated thr...         1.168921e-231

## Example of getting one answer

In [162]:
train_dataset['question'][0]

'Do nontriploid partial hydatidiform moles exist?'

In [197]:
train_dataset['long_answer'][0]

'Nontriploid partial moles probably do not exist: careful reevaluation of putative specimens will probably uncover pathologic or ploid errors in almost all cases.'

In [167]:
train_dataset['context'][0]['contexts']

['To study whether nontriploid partial hydatidiform moles truly exist.',
 'We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique.',
 'On review of the 19 moles, 53% (10/19) were diploid nonpartial moles (initially pathologically misclassified), and 37% (7/19) were triploid partial moles (initial ploidy misclassifications). One additional case (5%) was a diploid early complete mole (initially pathologically misclassified).']

In [175]:
train_dataset['context'][0]['contexts']

['To study whether nontriploid partial hydatidiform moles truly exist.',
 'We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique.',
 'On review of the 19 moles, 53% (10/19) were diploid nonpartial moles (initially pathologically misclassified), and 37% (7/19) were triploid partial moles (initial ploidy misclassifications). One additional case (5%) was a diploid early complete mole (initially pathologically misclassified).']

In [179]:
context_example = " ".join(train_dataset['context'][0]['contexts'])
context_example

'To study whether nontriploid partial hydatidiform moles truly exist. We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique. On review of the 19 moles, 53% (10/19) were diploid nonpartial moles (initially pathologically misclassified), and 37% (7/19) were triploid partial moles (initial ploidy misclassifications). One additional case (5%) was a diploid early complete mole (initially pathologically misclassified).'

In [180]:
import requests
import json

In [181]:
question = train_dataset['question'][0]
context = context_example

model_input = f"Answer the question based on the context below. " + \
    "\n Context: "  + context + \
    " Question: " + question
print("=================",model_input)
json_data = {
        "inputs": model_input,
        "parameters": {'temperature': 0.5,
                    'max_new_tokens': 100,
                    'return_full_text': False,
                    },
    }

# API_URL = "https://api-inference.huggingface.co/models/bigscience/bloom"
API_URL = "https://api-inference.huggingface.co/models/gpt2"
# API_URL = "https://api-inference.huggingface.co/models/openlm-research/open_llama_13b"
# API_URL = "https://api-inference.huggingface.co/models/EleutherAI/gpt-neox-20b"
# API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xxl"
# API_URL = "https://api-inference.huggingface.co/models/tiiuae/falcon-7b-instruct"
# API_URL = "https://api-inference.huggingface.co/models/conceptofmind/Hermes-Falcon-7b-4k"
headers = {"Authorization": "Bearer XXXXXXXXXXXXXXXX"} # ADD YOUR BEARER TOKEN

# HuggingFace response
response = requests.post(API_URL, headers=headers, json=json_data)
print(response)
json_response = json.loads(response.content.decode("utf-8"))
print(json_response)
model_output = json_response[0]['generated_text']
model_output

================= Answer the question based on the context below. 
 Context: To study whether nontriploid partial hydatidiform moles truly exist. We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique. On review of the 19 moles, 53% (10/19) were diploid nonpartial moles (initially pathologically misclassified), and 37% (7/19) were triploid partial moles (initial ploidy misclassifications). One additional case (5%) was a diploid early complete mole (initially pathologically misclassified). Question: Do nontriploid partial hydatidiform moles exist?
<Response [200]>
[{'generated_text': ' We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique. On review of the 19 moles, 53% (

' We conducted a reevaluation of pathology and ploidy in 19 putative nontriploid partial hydatidiform moles using standardized histologic diagnostic criteria and repeat flow cytometric testing by the Hedley technique. On review of the 19 moles, 53% (10/19) were diploid nonpartial moles (initially pathologically misclassified), and 37% (7/19) were triploid partial moles (initial ploidy misclassifications).'

## Evaluate Falcon-7B instruct dataset

In [189]:
import requests
import pandas as pd
import json

# API URL for the Hugging Face model
# API_URL = "https://api-inference.huggingface.co/models/gpt2"
# API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xxl"
API_URL = "https://api-inference.huggingface.co/models/tiiuae/falcon-7b-instruct"

# Create an empty list to store the results
results_falcon = []

# Iterate over each question, answer, and context
for idx, (question, answer, context) in enumerate(zip(questions, answers, contexts_list)):
#     print(question, answer)
    # Create the model input prompt
    model_input = f"Answer the question based on the context below.\nContext: {context}\nQuestion: {question}"
    headers = {"Authorization": "Bearer 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"}
#     print("=========model_input=========", model_input)
    # Define the payload for the API request
    payload = {
        "inputs": model_input,
        "parameters": {'temperature': 0.5,
                    'max_new_tokens': 50,
                    'return_full_text': False,
                    },
    }


    # HuggingFace response
    response = requests.post(API_URL, headers=headers, json=payload)
#     print(response)
    json_response = json.loads(response.content.decode("utf-8"))
    print("*********json_response**********",json_response)
    model_output = json_response[0]['generated_text']

    # Append the results to the list
    results_falcon.append({
        'Question': question,
        'Actual Answer': answer,
        'Generated Answer': model_output,
        'Context': context
    })

    # Print progress
    print(f"Processed {idx+1} / {len(questions)}")

# Create a Pandas DataFrame from the results
df_falcon = pd.DataFrame(results_falcon)

# Print the DataFrame
df_falcon.head()


*********json_response********** [{'generated_text': '\nAnswer: Yes, it can.'}]
Processed 1 / 200
*********json_response********** [{'generated_text': ' Answer: Yes.'}]
Processed 2 / 200
*********json_response********** [{'generated_text': ' Answer: Yes.'}]
Processed 3 / 200
*********json_response********** [{'generated_text': ''}]
Processed 4 / 200
*********json_response********** [{'generated_text': ' Answer: No, reperfusion is not required before index ischemia in order to achieve ischemic preconditioning.'}]
Processed 5 / 200
*********json_response********** [{'generated_text': ' Answer: Yes.'}]
Processed 6 / 200
*********json_response********** [{'generated_text': ''}]
Processed 7 / 200
*********json_response********** [{'generated_text': '\nAs per the context, the virus is a risk factor for the development of pterygium. The study was conducted on 100 pterygium specimens from Danish patients, and 20 normal conjunctival biopsy specimens. HPV was identified'}]
Processed 8 / 200
****

*********json_response********** [{'generated_text': ''}]
Processed 55 / 200
*********json_response********** [{'generated_text': '\nAnswer: Yes, home sampling kits for sexually transmitted infections are generally acceptable among MSM, and they offer potential benefits such as increased access, personal comfort, and empowerment. However, concerns regarding accuracy, delays in receiving results, and the possible negative impact'}]
Processed 56 / 200
*********json_response********** [{'generated_text': ''}]
Processed 57 / 200
*********json_response********** [{'generated_text': '\nBased on the provided context, it appears that EBP does not significantly improve patient outcomes compared to standard practice. However, EBP doctors attended twice as many patients/doctor as SP doctors, suggesting they may have better patient management skills.'}]
Processed 58 / 200
*********json_response********** [{'generated_text': ''}]
Processed 59 / 200
*********json_response********** [{'generated_text

*********json_response********** [{'generated_text': '\nAnswer: Yes, postoperative nomograms designed to predict BCR do a better job of predicting more clinically relevant prostate cancer outcomes than BCR.'}]
Processed 103 / 200
*********json_response********** [{'generated_text': '\nYes, carotid artery stenosis is a risk factor for perioperative stroke in patients undergoing coronary artery bypass surgery, and preoperative carotid screening is recommended for all patients.'}]
Processed 104 / 200
*********json_response********** [{'generated_text': '\nAnswer: Yes, according to the study, psychological distress is a predictor of disability due to common chronic disorders.'}]
Processed 105 / 200
*********json_response********** [{'generated_text': '\nAnswer: Yes, solid culture for tuberculosis can potentially influence clinical decision making in India.'}]
Processed 106 / 200
*********json_response********** [{'generated_text': '\nAnswer: Yes, elevated resting heart rate is a relevant m

*********json_response********** [{'generated_text': "\nThe question is asking whether the Department of Transportation's records and stroke survivors' self-reported data on motor vehicle collisions and driving convictions agree. The answer is that there is a significant level of agreement between the two sources of data, with k"}]
Processed 149 / 200
*********json_response********** [{'generated_text': ''}]
Processed 150 / 200
*********json_response********** [{'generated_text': ' Yes or No.\nAnswer: Yes.'}]
Processed 151 / 200
*********json_response********** [{'generated_text': ''}]
Processed 152 / 200
*********json_response********** [{'generated_text': ''}]
Processed 153 / 200
*********json_response********** [{'generated_text': ''}]
Processed 154 / 200
*********json_response********** [{'generated_text': '\nNo, the ADAM questionnaire is not useful for the screening of PADAM. PADAM can only be diagnosed by a blood test to measure serum testosterone levels.'}]
Processed 155 / 200
*

Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer  \
0  Our findings support CIMT as a significant ind...   
1  Comparison of mother-offspring and father-offs...   
2  Implant coating with an antibacterial-loaded h...   
3  This study that shows that adequate femoral tu...   
4  The preconditioning effect (infarct size reduc...   

                                    Generated Answer  \
0                             \nAnswer: Yes, it can.   
1                                       Answer: Yes.   
2                                       Answer: Yes.   
3                                                      
4   Answer: No, reperfusion is not required befor...   

                                             Context  
0  It is not known whether common carotid intima ...  
1  It has been suggested that increasing obesity ...  
2  Implant-related infections represent one of th...  
3  This study aims to study femoral tunnel length...  
4  Ischemic preconditioning (IP) is initiated thr...

In [191]:
from bert_score import score as bert_score
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge import Rouge
from sentence_transformers import SentenceTransformer
import numpy as np
import re
import string
import collections
from rouge_score import rouge_scorer

def normalize_answer(s):
  """Lower text and remove punctuation, articles and extra whitespace."""
  def remove_articles(text):
    regex = re.compile(r'\b(a|an|the)\b', re.UNICODE)
    return re.sub(regex, ' ', text)
  def white_space_fix(text):
    return ' '.join(text.split())
  def remove_punc(text):
    exclude = set(string.punctuation)
    return ''.join(ch for ch in text if ch not in exclude)
  def lower(text):
    return text.lower()
  return white_space_fix(remove_articles(remove_punc(lower(s))))

def get_tokens(s):
    if not s: return []
    return normalize_answer(s).split()

# Compute F1 score
def compute_f1(a_gold, a_pred):
    gold_toks = get_tokens(a_gold)
    pred_toks = get_tokens(a_pred)
    common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
    num_same = sum(common.values())
    if len(gold_toks) == 0 or len(pred_toks) == 0:
        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
        return int(gold_toks == pred_toks)
    if num_same == 0:
        return 0
    precision = 1.0 * num_same / len(pred_toks)
    recall = 1.0 * num_same / len(gold_toks)
    f1 = (2 * precision * recall) / (precision + recall)
    return f1

# Calculate F1 score
def calculate_f1_score(row):
    return compute_f1(row['Actual Answer'], row['Generated Answer'])

# Calculate BLEU score
def calculate_bleu_score(row):
    smoothie = SmoothingFunction().method4
    reference = row['Actual Answer'].split()
    hypothesis = row['Generated Answer'].split()
    return sentence_bleu([reference], hypothesis, smoothing_function=smoothie)

# Compute ROUGE score
def compute_rouge(answer, ideal_answer):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    # Calculate the ROUGE score
    score = scorer.score(answer, ideal_answer)
    # Extract the F1 score for ROUGE-1
    rouge_score = score['rouge1'].fmeasure
    return rouge_score

# Calculate ROUGE score
def calculate_rouge_score(row):
    return compute_rouge(row['Actual Answer'], row['Generated Answer'])

# Compute SentenceSim score
def sentence_similarity_alternate(ideal_answer, generated_answer):
    #Compute embedding for both lists
    embedding_1= model1.encode(ideal_answer, convert_to_tensor=True)
    embedding_2 = model1.encode(generated_answer, convert_to_tensor=True)
    sim_score = util.pytorch_cos_sim(embedding_1, embedding_2)
    sim_score = sim_score.numpy()[0][0]
    return sim_score

# Calculate SentenceSim score
def calculate_similarity_score(row):
    return sentence_similarity_alternate(row['Actual Answer'], row['Generated Answer'])

# Apply the functions to the DataFrame
df_falcon['F1 Score'] = df_falcon.apply(calculate_f1_score, axis=1)
df_falcon['BLEU Score'] = df_falcon.apply(calculate_bleu_score, axis=1)
df_falcon['ROUGE Score'] = df_falcon.apply(calculate_rouge_score, axis=1)
df_falcon['SentenceSim Score'] = df_falcon.apply(calculate_similarity_score, axis=1)

# Print the DataFrame with evaluation metrics
display(df_falcon.head())
df_falcon.describe()

Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer  \
0  Our findings support CIMT as a significant ind...   
1  Comparison of mother-offspring and father-offs...   
2  Implant coating with an antibacterial-loaded h...   
3  This study that shows that adequate femoral tu...   
4  The preconditioning effect (infarct size reduc...   

                                    Generated Answer  \
0                             \nAnswer: Yes, it can.   
1                                       Answer: Yes.   
2                                       Answer: Yes.   
3                                                      
4   Answer: No, reperfusion is not required befor...   

                                             Context  F1 Score  BLEU Score  \
0  It is not known whether common carotid intima ...  0.000000    0.000000   
1  It has been suggested that increasing obesity ...  0.000000    0.000000   
2  Implant-related infections represent one of th...  0.000000    0.000000   
3  This study aims to study femoral tunnel length...  0.000000    0.000000   
4  Ischemic preconditioning (IP) is initiated thr...  0.291667    0.005354   

   ROUGE Score  SentenceSim Score  
0     0.000000           0.033674  
1     0.000000          -0.006110  
2     0.000000           0.052875  
3     0.000000          -0.036668  
4     0.259259           0.725082

F1 Score  BLEU Score  ROUGE Score  SentenceSim Score
count  200.000000  200.000000   200.000000         200.000000
mean     0.147852    0.025753     0.162232           0.335198
std      0.169129    0.050103     0.182540           0.358516
min      0.000000    0.000000     0.000000          -0.130128
25%      0.000000    0.000000     0.000000           0.000296
50%      0.056548    0.000003     0.052201           0.093365
75%      0.286749    0.025565     0.322744           0.701171
max      0.680851    0.312469     0.733333           0.948647

#### Observation
Scores for the BLOOM model on 200 QnA test dataset:
- F-1 score: 0.147852
- BLEU score: 0.025753
- ROUGE Score: 0.162232	
- SentenceSim score: 0.335198

## Evaluate Redpajama model 


In [200]:
import requests
import pandas as pd
import json

# API URL for the Hugging Face model
API_URL = "https://api-inference.huggingface.co/models/togethercomputer/RedPajama-INCITE-Base-3B-v1"

# Create an empty list to store the results
results_redpajama = []

# Iterate over each question, answer, and context
for idx, (question, answer, context) in enumerate(zip(questions, answers, contexts_list)):
#     print(question, answer)
    # Create the model input prompt
    model_input = f"Answer the question based on the context below.\nContext: {context}\nQuestion: {question}"
    headers = {"Authorization": "Bearer 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"}
#     print("=========model_input=========", model_input)
    # Define the payload for the API request
    payload = {
        "inputs": model_input,
        "parameters": {'temperature': 0.5,
                    'max_new_tokens': 50,
                    'return_full_text': False,
                    },
    }


    # HuggingFace response
    response = requests.post(API_URL, headers=headers, json=payload)
#     print(response)
    json_response = json.loads(response.content.decode("utf-8"))
    print("*********json_response**********",json_response)
    model_output = json_response[0]['generated_text']

    # Append the results to the list
    results_redpajama.append({
        'Question': question,
        'Actual Answer': answer,
        'Generated Answer': model_output,
        'Context': context
    })

    # Print progress
    print(f"Processed {idx+1} / {len(questions)}")

# Create a Pandas DataFrame from the results
df_redpajama = pd.DataFrame(results_redpajama)

# Print the DataFrame
df_redpajama.head()


*********json_response********** [{'generated_text': '\nContext: The aim of this study was to determine the prevalence of and risk factors for hypertension in a rural community in the Philippines.\nContext: The prevalence of hypertension in the Philippines is high, and the prevalence of hypertension in the rural areas is'}]
Processed 1 / 200
*********json_response********** [{'generated_text': '\nContext: The association between maternal body mass index (BMI) and offspring BMI has been investigated in many studies. However, the strength of the association between maternal BMI and offspring BMI has not been investigated in a large population-based study.\nQuestion'}]
Processed 2 / 200
*********json_response********** [{'generated_text': '\nAnswer: Yes.\nQuestion: Is intraoperative coating of implants with antibacterial-loaded hydrogel feasible and resistant to press-fit insertion?\nAnswer: Yes. The coating was completely released from the implant after press-fit insertion.\nQuestion: Ca

*********json_response********** [{'generated_text': '\nAnswer: No.\nThe study was not designed to answer this question.\nThe study was not designed to answer this question. The study was not designed to answer this question.\nThe study was not designed to answer this question. The study'}]
Processed 27 / 200
*********json_response********** [{'generated_text': '\nContext: The prevalence of depression is higher in ethnic minority groups than in the white British population. However, the relationship between neighbourhood ethnic density and depression diagnosis and prescribing is not well understood.\nQuestion: Does depression diagnosis and antidepressant prescribing vary by'}]
Processed 28 / 200
*********json_response********** [{'generated_text': '\nAnswer: No.\nThe authors concluded that there was no difference in the 3-month outcome between men and women treated with intravenous thrombolysis for acute ischemic stroke.\nThe authors stated that the results of this study should be inte

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /var/folders/gy/6lyfn5752472j2htx4g01gc40000gn/T/ipykernel_40154/3357684751.py:33 in <module>    │
│                                                                                                  │
│ [Errno 2] No such file or directory:                                                             │
│ '/var/folders/gy/6lyfn5752472j2htx4g01gc40000gn/T/ipykernel_40154/3357684751.py'                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 0

In [ ]:
from bert_score import score as bert_score
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge import Rouge
from sentence_transformers import SentenceTransformer
import numpy as np
import re
import string
import collections
from rouge_score import rouge_scorer

def normalize_answer(s):
  """Lower text and remove punctuation, articles and extra whitespace."""
  def remove_articles(text):
    regex = re.compile(r'\b(a|an|the)\b', re.UNICODE)
    return re.sub(regex, ' ', text)
  def white_space_fix(text):
    return ' '.join(text.split())
  def remove_punc(text):
    exclude = set(string.punctuation)
    return ''.join(ch for ch in text if ch not in exclude)
  def lower(text):
    return text.lower()
  return white_space_fix(remove_articles(remove_punc(lower(s))))

def get_tokens(s):
    if not s: return []
    return normalize_answer(s).split()

# Compute F1 score
def compute_f1(a_gold, a_pred):
    gold_toks = get_tokens(a_gold)
    pred_toks = get_tokens(a_pred)
    common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
    num_same = sum(common.values())
    if len(gold_toks) == 0 or len(pred_toks) == 0:
        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
        return int(gold_toks == pred_toks)
    if num_same == 0:
        return 0
    precision = 1.0 * num_same / len(pred_toks)
    recall = 1.0 * num_same / len(gold_toks)
    f1 = (2 * precision * recall) / (precision + recall)
    return f1

# Calculate F1 score
def calculate_f1_score(row):
    return compute_f1(row['Actual Answer'], row['Generated Answer'])

# Calculate BLEU score
def calculate_bleu_score(row):
    smoothie = SmoothingFunction().method4
    reference = row['Actual Answer'].split()
    hypothesis = row['Generated Answer'].split()
    return sentence_bleu([reference], hypothesis, smoothing_function=smoothie)

# Compute ROUGE score
def compute_rouge(answer, ideal_answer):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    # Calculate the ROUGE score
    score = scorer.score(answer, ideal_answer)
    # Extract the F1 score for ROUGE-1
    rouge_score = score['rouge1'].fmeasure
    return rouge_score

# Calculate ROUGE score
def calculate_rouge_score(row):
    return compute_rouge(row['Actual Answer'], row['Generated Answer'])

# Compute SentenceSim score
def sentence_similarity_alternate(ideal_answer, generated_answer):
    #Compute embedding for both lists
    embedding_1= model1.encode(ideal_answer, convert_to_tensor=True)
    embedding_2 = model1.encode(generated_answer, convert_to_tensor=True)
    sim_score = util.pytorch_cos_sim(embedding_1, embedding_2)
    sim_score = sim_score.numpy()[0][0]
    return sim_score

# Calculate SentenceSim score
def calculate_similarity_score(row):
    return sentence_similarity_alternate(row['Actual Answer'], row['Generated Answer'])

# Apply the functions to the DataFrame
df_redpajama['F1 Score'] = df_redpajama.apply(calculate_f1_score, axis=1)
df_redpajama['BLEU Score'] = df_redpajama.apply(calculate_bleu_score, axis=1)
df_redpajama['ROUGE Score'] = df_redpajama.apply(calculate_rouge_score, axis=1)
df_redpajama['SentenceSim Score'] = df_redpajama.apply(calculate_similarity_score, axis=1)

# Print the DataFrame with evaluation metrics
display(df_redpajama.head())
df_redpajama.describe()

In [204]:
import requests
import pandas as pd
import json

# API URL for the Hugging Face model
API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xxl"
# API_URL = "https://api-inference.huggingface.co/models/tiiuae/falcon-7b-instruct"

# Create an empty list to store the results
results_flan_t5 = []

# Iterate over each question, answer, and context
for idx, (question, answer, context) in enumerate(zip(questions, answers, contexts_list)):
#     print(question, answer)
    # Create the model input prompt
    model_input = f"Answer the question based on the context below.\nContext: {context}\nQuestion: {question}"
    headers = {"Authorization": "Bearer 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"}
#     print("=========model_input=========", model_input)
    # Define the payload for the API request
    payload = {
        "inputs": model_input,
        "parameters": {'temperature': 0.5,
                    'max_new_tokens': 50,
                    'return_full_text': False,
                    },
    }


    # HuggingFace response
    response = requests.post(API_URL, headers=headers, json=payload)
#     print(response)
    json_response = json.loads(response.content.decode("utf-8"))
    print("*********json_response**********",json_response)
    model_output = json_response[0]['generated_text']

    # Append the results to the list
    results_flan_t5.append({
        'Question': question,
        'Actual Answer': answer,
        'Generated Answer': model_output,
        'Context': context
    })

    # Print progress
    print(f"Processed {idx+1} / {len(questions)}")

# Create a Pandas DataFrame from the results
df_flant5 = pd.DataFrame(results_flan_t5)

# Print the DataFrame
df_flant5.head()


*********json_response********** [{'generated_text': 'yes'}]
Processed 1 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 2 / 200
*********json_response********** [{'generated_text': 'Yes'}]
Processed 3 / 200
*********json_response********** [{'generated_text': 'Yes'}]
Processed 4 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 5 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 6 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 7 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 8 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 9 / 200
*********json_response********** [{'generated_text': 'Socioeconomic differences in mortality persist after retirement.'}]
Processed 10 / 200
*********json_response********** [{'generated_text': 'No'}]
Processed 11 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 12 

*********json_response********** [{'generated_text': 'Yes'}]
Processed 83 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 84 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 85 / 200
*********json_response********** [{'generated_text': 'No'}]
Processed 86 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 87 / 200
*********json_response********** [{'generated_text': 'Longer LOS in the ICU and non-ICU LOS>10 days, and HO prophylaxis with XRT were significantly associated with the development of HO. Other variables evaluated in multivariate modeling not'}]
Processed 88 / 200
*********json_response********** [{'generated_text': 'Yes'}]
Processed 89 / 200
*********json_response********** [{'generated_text': 'Obesity has a negative impact on the pattern and outcome of trauma in children.'}]
Processed 90 / 200
*********json_response********** [{'generated_text': 'Covering of the resection margins after distal p

*********json_response********** [{'generated_text': 'no'}]
Processed 164 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 165 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 166 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 167 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 168 / 200
*********json_response********** [{'generated_text': 'The results of this study suggest that the failure of ACLR is associated with steep posterior tibial slopes.'}]
Processed 169 / 200
*********json_response********** [{'generated_text': 'yes'}]
Processed 170 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 171 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 172 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 173 / 200
*********json_response********** [{'generated_text': 'no'}]
Processed 174 / 200
*********jso

Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer Generated Answer  \
0  Our findings support CIMT as a significant ind...              yes   
1  Comparison of mother-offspring and father-offs...               no   
2  Implant coating with an antibacterial-loaded h...              Yes   
3  This study that shows that adequate femoral tu...              Yes   
4  The preconditioning effect (infarct size reduc...               no   

                                             Context  
0  It is not known whether common carotid intima ...  
1  It has been suggested that increasing obesity ...  
2  Implant-related infections represent one of th...  
3  This study aims to study femoral tunnel length...  
4  Ischemic preconditioning (IP) is initiated thr...

In [205]:
from bert_score import score as bert_score
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge import Rouge
from sentence_transformers import SentenceTransformer
import numpy as np
import re
import string
import collections
from rouge_score import rouge_scorer

def normalize_answer(s):
  """Lower text and remove punctuation, articles and extra whitespace."""
  def remove_articles(text):
    regex = re.compile(r'\b(a|an|the)\b', re.UNICODE)
    return re.sub(regex, ' ', text)
  def white_space_fix(text):
    return ' '.join(text.split())
  def remove_punc(text):
    exclude = set(string.punctuation)
    return ''.join(ch for ch in text if ch not in exclude)
  def lower(text):
    return text.lower()
  return white_space_fix(remove_articles(remove_punc(lower(s))))

def get_tokens(s):
    if not s: return []
    return normalize_answer(s).split()

# Compute F1 score
def compute_f1(a_gold, a_pred):
    gold_toks = get_tokens(a_gold)
    pred_toks = get_tokens(a_pred)
    common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
    num_same = sum(common.values())
    if len(gold_toks) == 0 or len(pred_toks) == 0:
        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
        return int(gold_toks == pred_toks)
    if num_same == 0:
        return 0
    precision = 1.0 * num_same / len(pred_toks)
    recall = 1.0 * num_same / len(gold_toks)
    f1 = (2 * precision * recall) / (precision + recall)
    return f1

# Calculate F1 score
def calculate_f1_score(row):
    return compute_f1(row['Actual Answer'], row['Generated Answer'])

# Calculate BLEU score
def calculate_bleu_score(row):
    smoothie = SmoothingFunction().method4
    reference = row['Actual Answer'].split()
    hypothesis = row['Generated Answer'].split()
    return sentence_bleu([reference], hypothesis, smoothing_function=smoothie)

# Compute ROUGE score
def compute_rouge(answer, ideal_answer):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    # Calculate the ROUGE score
    score = scorer.score(answer, ideal_answer)
    # Extract the F1 score for ROUGE-1
    rouge_score = score['rouge1'].fmeasure
    return rouge_score

# Calculate ROUGE score
def calculate_rouge_score(row):
    return compute_rouge(row['Actual Answer'], row['Generated Answer'])

# Compute SentenceSim score
def sentence_similarity_alternate(ideal_answer, generated_answer):
    #Compute embedding for both lists
    embedding_1= model1.encode(ideal_answer, convert_to_tensor=True)
    embedding_2 = model1.encode(generated_answer, convert_to_tensor=True)
    sim_score = util.pytorch_cos_sim(embedding_1, embedding_2)
    sim_score = sim_score.numpy()[0][0]
    return sim_score

# Calculate SentenceSim score
def calculate_similarity_score(row):
    return sentence_similarity_alternate(row['Actual Answer'], row['Generated Answer'])

# Apply the functions to the DataFrame
df_flant5['F1 Score'] = df_flant5.apply(calculate_f1_score, axis=1)
df_flant5['BLEU Score'] = df_flant5.apply(calculate_bleu_score, axis=1)
df_flant5['ROUGE Score'] = df_flant5.apply(calculate_rouge_score, axis=1)
df_flant5['SentenceSim Score'] = df_flant5.apply(calculate_similarity_score, axis=1)

# Print the DataFrame with evaluation metrics
df_flant5.to_csv('df_flan_t5_pretrained_eval.csv')
display(df_flant5.head())
df_flant5.describe()


Question  \
0  Can common carotid intima media thickness serv...   
1     Is there an intrauterine influence on obesity?   
2  Does implant coating with antibacterial-loaded...   
3  Can Flexible Instruments Create Adequate Femor...   
4  Does ischemic preconditioning require reperfus...   

                                       Actual Answer Generated Answer  \
0  Our findings support CIMT as a significant ind...              yes   
1  Comparison of mother-offspring and father-offs...               no   
2  Implant coating with an antibacterial-loaded h...              Yes   
3  This study that shows that adequate femoral tu...              Yes   
4  The preconditioning effect (infarct size reduc...               no   

                                             Context  F1 Score  BLEU Score  \
0  It is not known whether common carotid intima ...       0.0         0.0   
1  It has been suggested that increasing obesity ...       0.0         0.0   
2  Implant-related infections represent one of th...       0.0         0.0   
3  This study aims to study femoral tunnel length...       0.0         0.0   
4  Ischemic preconditioning (IP) is initiated thr...       0.0         0.0   

   ROUGE Score  SentenceSim Score  
0          0.0          -0.027220  
1          0.0           0.067637  
2          0.0          -0.023249  
3          0.0          -0.039807  
4          0.0          -0.014294

F1 Score  BLEU Score  ROUGE Score  SentenceSim Score
count  200.000000  200.000000   200.000000         200.000000
mean     0.063637    0.006748     0.067779           0.152722
std      0.131048    0.022264     0.139672           0.288905
min      0.000000    0.000000     0.000000          -0.114598
25%      0.000000    0.000000     0.000000          -0.016939
50%      0.000000    0.000000     0.000000           0.023571
75%      0.000000    0.000000     0.000000           0.093360
max      0.545455    0.183504     0.636364           0.930428